util/gen_doc_hw_summary_table.py

```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""Collect the one-line summaries of all hardware blocks and output them as
+Markdown-formatted table in the README.md at the top of the hardware block tree.
+"""
+
+from dataclasses import dataclass
+import hjson
+import logging as log
+import os
+import pathlib
+from typing import List
+import unittest
+
+
+def main():
+    this_file_path = pathlib.Path(__file__)
+    repo_root = this_file_path.parent.parent.resolve()
+
+    # Collect entries for the HW summary from directories under hw/ip.
+    hw_ip_dir = repo_root / 'hw' / 'ip'
+    summary_entries = []
+    for p in os.scandir(hw_ip_dir):
+        if not p.is_dir():
+            continue
+
+        # Load data file, and skip blocks that don't provide one.
+        name = p.name
+        data_path = pathlib.Path(p) / 'data' / f'{name}.hjson'
+        if not data_path.exists():
+            log.warning(f'Skipping `{name}` because data file not found.')
+            continue
+        with open(data_path, 'r') as f:
+            data = hjson.loads(f.read())
+
+        name = data['name']
+
+        # Get one-line description, and skip blocks that don't have one.
+        if 'one_line_desc' not in data:
+            log.warning(f'Skipping `{name}` because no `one_line_desc` found '
+                        'in its data file.')
+            continue
+        one_liner = data['one_line_desc']
+
+        # Create HW summary entry and append it to our list.
+        summary_entries.append(HwSummaryEntry(name, one_liner))
+
+    # Create table summarizing HW blocks.
+    table = HwSummaryTable(summary_entries)
+
+    # Write table into hw/ip/README.md, replacing any content that was
+    # previously in the section marked as autogenerated by this tool.
+    with open(hw_ip_dir / 'README.md', 'r') as f:
+        text = f.read()
+        new_section = (
+            '<!-- Do NOT edit this section manually, '
+            f'it is generated by {os.path.relpath(this_file_path, repo_root)}. '
+            '-->\n')
+        new_section += table.to_markdown().rstrip('\n')
+        new_text = replace_section_in_text(
+            text,
+            '\n<!-- BEGIN AUTOGEN Summary Table -->\n',
+            '\n<!-- END AUTOGEN Summary Table -->\n',
+            new_section)
+    with open(hw_ip_dir / 'README.md', 'w') as f:
+        f.write(new_text)
+
+
+@dataclass
+class HwSummaryEntry:
+    """Simple dataclass summarizing one HW block."""
+    block_name: str
+    one_line_desc: str
+
+    def name_as_linked_markdown(self) -> str:
+        return f'[{self.block_name}]({self.block_name})'
```

Comment:
```suggestion
        return f'[{self.block_name}]({self.block_name}/README.md)'
```

Pointing to the actual file would be preferable.

Reply:
Hmm, are you sure?
- The table is intentionally readable also as plain text, even though it's quite wide. This would shift each description further right by 10 more characters, which add little value to the reader.
- When rendered on GitHub, clicking the link currently takes you to the folder of the HW IP block where its README is displayed at the bottom, but you can also decide to continue to code or DV or other documentation. Linking directly to the README would put the code to the background.
- When rendered by mdbook, linking to `<block_name>` rather than `<block_name>/README.md` is sufficient as well, right?

Why do you think pointing to each README is preferable?

Reply:
If we'd like the readable table and linking straight to the `README` file, we could use Markdown's footnote style links:

```markdown
| HW Block    | Brief Summary                                                                        |
|-------------|--------------------------------------------------------------------------------------|
| [adc_ctrl]  | Low-power controller for a dual-channel ADC with filtering and debouncing capability |
| [aes]       | AES encryption and decryption engine with SCA and FI countermeasures                 |
| [aon_timer] | Wakeup and watchdog timers running on a low-power, always-on clock                   |
...

[adc_ctrl]: ./adc_ctrl/README.md
[aes]: ./aes/README.md
[aon_timer]: ./aon_timer/README.md
...
```

We do this for checklists:
<https://raw.githubusercontent.com/lowRISC/opentitan/master/hw/top_earlgrey/ip_autogen/rv_plic/doc/checklist.md>

Reply:
I would like there to be a distinction between linking to a folder and to a file. I think this is semantically clearer, but also can have some practical benefits:
* It would allow us to substitute links to folders and files that are not rendered in the documentation with their github URL. Something, I'd like to do in the near future to make links to source files and folders work well in the rendered docs.
* Linking to files directly means that if anchors are added to the link, standard link checkers can unambiguously check that these anchors are valid.

I have no strong opinion on the link style.

Reply:
Thanks for the explanation, @HU90m. I understand those points and will change the links to point to the respective `README.md`.

Thanks for the suggestion, @jwnt. I'll give shortcut reference style links a try if they aren't overly complex to generate.

Reply:
Done